### PR TITLE
[SPARK-38423][K8S] Reuse driver pod's `priorityClassName` for `PodGroup`

### PR DESCRIPTION
**Pull request by Yikun**
### What changes were proposed in this pull request?
This patch set podgroup `priorityClassName` to `driver.pod.spec.priorityClassName`.

### Why are the changes needed?
Support priority scheduling with Volcano implementations

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
- New UT to make sure feature step set podgroup priority as expected.
- Add two integration tests:
  - 1. Submit 3 different priority jobs (spark pi) to make sure job completed result as expected.
  - 2. Submit 3 different priority jobs (driver submisson) to make sure job scheduler order as expected.
- All existing UT and IT

### CODE REVIEW

File: docs/running-on-kubernetes.md
Comment (martin-g) [inline docs/running-on-kubernetes.md:1363]: `queue` ? looks like a copy/paste error

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala
Comment (martin-g) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/Config.scala:296]: again the doc talks about `queue` (https://volcano.sh/en/docs/vcjob/#queue) while it should be https://volcano.sh/en/docs/vcjob/#priorityclassname

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I'm wondering why we need all these trivial static configurations one by one.
Apache Spark already has `spark.kubernetes.driver.podTemplateFile`. That's the official way which Apache Spark support `  priorityClassName`.

**Comment by holdenk** [comment]
If we don't merge this (which I think yeah pod template is a reasonable alternative), let's document how to set the priority class with the pod template.

**Comment by dongjoon-hyun** [comment]
> If we don't merge this (which I think yeah pod template is a reasonable alternative), let's document how to set the priority class with the pod template.

+1 for @holdenk 's suggestion.

**Comment by Yikun** [comment]
@holdenk @dongjoon-hyun OK, it's also fine for me, I'd like to add a guide doc to help users enable priority scheduling. I will add soon.

**Comment by dongjoon-hyun** [comment]
You can reuse SPARK-38189 by converting into a documentation PR, @Yikun .

**Comment by dongjoon-hyun** [comment]
To help you make your Volcano PR, I made a `Pod`-spec `priorityClassName` test case PR independently.
- https://github.com/apache/spark/pull/35716

**Comment by Yikun** [comment]
@dongjoon-hyun Thanks, your PR can make sure priority class name work basiclly.

 I also drafted a[ multi-job IT](https://github.com/Yikun/spark/commit/675f56b8ba97c22c1d572e6b07501c5af3b8607e#diff-e8b4a160e455ae14cdc70e658c40e870c29a880f40681f1bb463a61efdb5f242R230) before. Do you think we need this? TLDR: First, use a pod to stuck all jobs, then to check all jobs scheduled time order by priority value.

**Comment by dongjoon-hyun** [comment]
+1 for adding those tests for `Volcano` module, @Yikun .
> // Prepare the priority resource
>  createOrReplaceYAMLResource(VOLCANO_PRIORITY_YAML)

**Comment by dongjoon-hyun** [comment]
Please check the dev mailing list . Ramping down for Apache Spark 3.3 release started, @Yikun .
- https://lists.apache.org/thread/ffdk52hkvgsc5ncjh1z0nv120jowtrld

**Comment by Yikun** [comment]
Replace by https://github.com/apache/spark/pull/35728

**Comment by dongjoon-hyun** [comment]
Well, I don't think #35728 replaces this. You need write a document for `Volcano` pod scheduling with volcano priority.

**Comment by Yikun** [comment]
@dongjoon-hyun Got it

**Comment by dongjoon-hyun** [comment]
I merged https://github.com/apache/spark/pull/35728 to master/branch-3.2 for Apache Spark 3.2.2.

**Comment by Yikun** [comment]
@dongjoon-hyun @holdenk 

- Volcano support Pod priority scheduling by default (by creating a temp podgroup in volcano internally), we can only set schedulerName to `vocano` and don't specified feature step. 
- (This PR) If we want to support the priority with collaborative work with queue/minRes(comming soon). We need to interit the driver pod priorityClassName as job(podgroup) priority.

**Comment by Yikun** [comment]
```
[info] VolcanoSuite:
[info] - Run SparkPi with no resources (13 seconds, 181 milliseconds)
[info] - Run SparkPi with no resources & statefulset allocation (12 seconds, 772 milliseconds)
[info] - Run SparkPi with a very long application name. (11 seconds, 736 milliseconds)
[info] - Use SparkLauncher.NO_RESOURCE (12 seconds, 821 milliseconds)
[info] - Run SparkPi with a master URL without a scheme. (11 seconds, 743 milliseconds)
[info] - Run SparkPi with an argument. (13 seconds, 714 milliseconds)
[info] - Run SparkPi with custom labels, annotations, and environment variables. (11 seconds, 781 milliseconds)                            
[info] - All pods have the same service account by default (11 seconds, 737 milliseconds)
[info] - Run extraJVMOptions check on driver (5 seconds, 702 milliseconds)
[info] - Run SparkRemoteFileTest using a remote data file (12 seconds, 912 milliseconds)
[info] - Verify logging configuration is picked from the provided SPARK_CONF_DIR/log4j2.properties (20 seconds, 239 milliseconds)          
[info] - Run SparkPi with env and mount secrets. (20 seconds, 863 milliseconds)
[info] - Run PySpark on simple pi.py example (13 seconds, 795 milliseconds)
[info] - Run PySpark to test a pyfiles example (15 seconds, 894 milliseconds)
[info] - Run PySpark with memory customization (13 seconds, 816 milliseconds)
[info] - Run in client mode. (8 seconds, 186 milliseconds)
[info] - Start pod creation from template (11 seconds, 870 milliseconds)
[info] - SPARK-38398: Schedule pod creation from template (12 seconds, 814 milliseconds)
[info] - Test basic decommissioning (47 seconds, 163 milliseconds)
[info] - Test basic decommissioning with shuffle cleanup (47 seconds, 300 milliseconds)
[info] - Test decommissioning with dynamic allocation & shuffle cleanups (2 minutes, 46 seconds)
[info] - Test decommissioning timeouts (48 seconds, 461 milliseconds)
[info] - SPARK-37576: Rolling decommissioning (1 minute, 8 seconds)
[info] - Run SparkPi with volcano scheduler (11 seconds, 852 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (only 1 enable) (17 seconds, 300 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (all enable) (34 seconds, 525 milliseconds)
[info] - SPARK-38189: Run SparkPi Jobs with priorityClassName (21 seconds, 146 milliseconds)
[info] - SPARK-38189: Run driver job to validate priority order (19 seconds, 278 milliseconds)
```

File: resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala
Comment (Yikun) [inline resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/k8s/features/VolcanoFeatureStep.scala:54]: Note: here is the core modification. Besides this file, all left files are for test.

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/VolcanoTestsSuite.scala
Comment (viirya) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/VolcanoTestsSuite.scala:332]: Hm, where did you use `LOW_PRIORITY_DRIVER_TEMPLATE`, `MEDIUM_PRIORITY_DRIVER_TEMPLATE` and `HIGH_PRIORITY_DRIVER_TEMPLATE`?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/VolcanoTestsSuite.scala
Comment (viirya) [inline resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/VolcanoTestsSuite.scala:307]: What are the two "TODO" for? Do you need to add other JIRA for it?

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/VolcanoTestsSuite.scala
Comment (Yikun) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/VolcanoTestsSuite.scala:332]: ah, this can be cleanup now, will address

File: resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/VolcanoTestsSuite.scala
Comment (Yikun) [reply resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/k8s/integrationtest/VolcanoTestsSuite.scala:307]: after this merged: https://github.com/apache/spark/pull/35733

we can cleanup this yaml.

**Comment by Yikun** [comment]
@viirya Comments addressed.

**Comment by Yikun** [comment]
```
[info] VolcanoSuite:
[info] - Run SparkPi with volcano scheduler (12 seconds, 238 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (only 1 enable) (18 seconds, 354 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (all enable) (38 seconds, 685 milliseconds)
[info] - SPARK-38423: Run SparkPi Jobs with priorityClassName (21 seconds, 293 milliseconds)
[info] - SPARK-38423: Run driver job to validate priority order (19 seconds, 304 milliseconds)
[info] Run completed in 1 minute, 53 seconds.
[info] Total number of tests run: 5
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 5, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 152 s (02:32), completed Mar 6, 2022 3:58:34 PM
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
.

**Review by dongjoon-hyun — COMMENTED** [review]
Thank you for updates. I'm reviewing again.

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM (Pending rebasing), @Yikun .

**Review by viirya — APPROVED** [review]

**Comment by dongjoon-hyun** [comment]
#35733 is merged now . Could you rebase this PR, @Yikun ?

**Comment by Yikun** [comment]
```
[info] VolcanoSuite:
[info] - Run SparkPi with volcano scheduler (13 seconds, 276 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (only 1 enabled) (18 seconds, 289 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (all enabled) (27 seconds, 377 milliseconds)
[info] - SPARK-38423: Run SparkPi Jobs with priorityClassName (21 seconds, 209 milliseconds)
[info] - SPARK-38423: Run driver job to validate priority order (19 seconds, 212 milliseconds)
[info] Run completed in 1 minute, 43 seconds.
[info] Total number of tests run: 5
[info] Suites: completed 2, aborted 0
[info] Tests: succeeded 5, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 321 s (05:21), completed Mar 7, 2022 3:04:23 PM
```
Especially, check the rebase diff work.
```
$ build/sbt -Pvolcano -Pkubernetes -Pkubernetes-integration-tests -Dtest.exclude.tags=minikube,r  -Dtest.include.tags=volcano  -Dspark.kubernetes.test.namespace=default "kubernetes-integration-tests/test"

$ k get pod
No resources found in default namespace.

$ k get queue
NAME      AGE
default   2d19h
```

**Comment by Yikun** [comment]
@dongjoon-hyun @viirya Thanks for your review, rebased.

**Review by dongjoon-hyun — APPROVED** [review]
Thanks for rebasing. I also verified manually that all `queue`, `priorityClass`, `podGroup` are created and used and deleted correctly.
```
$ build/sbt -Psparkr -Pkubernetes -Pvolcano -Pkubernetes-integration-tests -Dtest.exclude.tags=minikube -Dspark.kubernetes.test.deployMode=docker-for-desktop "kubernetes-integration-tests/testOnly *.VolcanoSuite"
...
[info] VolcanoSuite:
[info] - Run SparkPi with no resources (9 seconds, 827 milliseconds)
[info] - Run SparkPi with no resources & statefulset allocation (9 seconds, 670 milliseconds)
[info] - Run SparkPi with a very long application name. (10 seconds, 666 milliseconds)
[info] - Use SparkLauncher.NO_RESOURCE (10 seconds, 649 milliseconds)
[info] - Run SparkPi with a master URL without a scheme. (10 seconds, 641 milliseconds)
[info] - Run SparkPi with an argument. (10 seconds, 645 milliseconds)
[info] - Run SparkPi with custom labels, annotations, and environment variables. (9 seconds, 619 milliseconds)
[info] - All pods have the same service account by default (10 seconds, 698 milliseconds)
[info] - Run extraJVMOptions check on driver (5 seconds, 619 milliseconds)
[info] - Run SparkRemoteFileTest using a remote data file (10 seconds, 729 milliseconds)
[info] - Verify logging configuration is picked from the provided SPARK_CONF_DIR/log4j2.properties (15 seconds, 113 milliseconds)
[info] - Run SparkPi with env and mount secrets. (20 seconds, 771 milliseconds)
[info] - Run PySpark on simple pi.py example (11 seconds, 681 milliseconds)
[info] - Run PySpark to test a pyfiles example (12 seconds, 729 milliseconds)
[info] - Run PySpark with memory customization (10 seconds, 625 milliseconds)
[info] - Run in client mode. (7 seconds, 307 milliseconds)
[info] - Start pod creation from template (10 seconds, 794 milliseconds)
[info] - SPARK-38398: Schedule pod creation from template (10 seconds, 739 milliseconds)
[info] - Test basic decommissioning (42 seconds, 10 milliseconds)
[info] - Test basic decommissioning with shuffle cleanup (43 seconds, 153 milliseconds)
[info] *** Test still running after 2 minutes, 20 seconds: suite name: VolcanoSuite, test name: Test decommissioning with dynamic allocation & shuffle cleanups.
[info] - Test decommissioning with dynamic allocation & shuffle cleanups (2 minutes, 42 seconds)
[info] - Test decommissioning timeouts (42 seconds, 740 milliseconds)
[info] - SPARK-37576: Rolling decommissioning (1 minute, 7 seconds)
[info] - Run SparkR on simple dataframe.R example (12 seconds, 694 milliseconds)
[info] - Run SparkPi with volcano scheduler (10 seconds, 655 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (only 1 enabled) (13 seconds, 282 milliseconds)
[info] - SPARK-38188: Run SparkPi jobs with 2 queues (all enabled) (19 seconds, 295 milliseconds)
[info] - SPARK-38423: Run SparkPi Jobs with priorityClassName (13 seconds, 181 milliseconds)
[info] - SPARK-38423: Run driver job to validate priority order (16 seconds, 439 milliseconds)
[info] Run completed in 15 minutes, 18 seconds.
[info] Total number of tests run: 29
[info] Suites: completed 1, aborted 0
[info] Tests: succeeded 29, failed 0, canceled 0, ignored 0, pending 0
[info] All tests passed.
[success] Total time: 942 s (15:42), completed Mar 6, 2022 11:53:04 PM
```

**Comment by dongjoon-hyun** [comment]
Merged to master for Apache Spark 3.3.0.

**Pull request closed without merging**
